packages/microcosm/src/agent.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * @fileoverview An agent is a general abstraction for intercepting actions.
+ * It is the basis of Domain and Effect
+ * @flow
+ */
+
+import { type Microcosm } from '../src/microcosm'
+import { Subject } from './subject'
+import { merge } from './data'
+
+export class Agent extends Subject {
+  repo: Microcosm
+  options: Object
+
+  static defaults: ?Object
+
+  constructor(repo: Microcosm, options?: Object) {
+    super(null, options)
+
+    this.repo = repo
+    this.options = merge(repo.options, this.constructor.defaults, options)
+
+    let tracker = repo.history.subscribe(this.receive.bind(this))
+
+    repo.subscribe({
+      complete: () => {
+        tracker.unsubscribe()
+        this.teardown(this.repo, this.options)
+      }
```

Comment:
looks like this is existing functionality from domains/effects, but what would cause a `complete` action?

Reply:
Microcosm's are Observables (implemented as Subjects ). When you call `complete()` on a Microcosm, it has the same effect as calling `shutdown` in Microcosm 12. Instead of subscribing to the `teardown` event using an event emitter, repos/domains/presenters now listen to the `complete` lifecycle event.

TBH, I think there's an improvement that could be made here. This feels like bad design. I wonder if I should give Subject itself a setup and teardown method. You wouldn't have to implement it, but it could be inherited through Agent to Domains and Effects.